viewing_party/party.py

```diff
@@ -1,23 +1,116 @@
+    else:
+        new_movie["title"] = title
+        new_movie["genre"] = genre
+        new_movie["rating"] = rating
+        return new_movie
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if title != movie["title"]:
+            continue
+        elif movie["title"] == title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+            return user_data
+        
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+    if not len(user_data["watched"]):
+        return 0.0
+    ratings = [movie["rating"] for movie in user_data["watched"]]
+    return statistics.mean(ratings)
+
+def get_most_watched_genre(user_data):
+    if not len(user_data["watched"]):
+        return None
+    genres = [movie["genre"]for movie in user_data["watched"]]
+    return mode(genres)
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    user_movies = get_user_movies(user_data)
+    friend_movies = get_friends_movies(user_data)
+    return [movie for movie in user_movies if movie not in friend_movies]
+
+def get_friends_unique_watched(user_data):
+    user_movies = get_user_movies(user_data)
+    friend_movies = get_friends_movies(user_data)
+    unique_movies = []
+    for movie in friend_movies:
+        if movie not in user_movies and movie not in unique_movies:
+            unique_movies.append(movie)
+    return unique_movies
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    recommended = []
+    friends_watched = get_friends_unique_watched(user_data)
+    for movie in friends_watched:
+        if movie not in user_data["watched"]:
+            if movie["host"] in user_data["subscriptions"] and movie not in recommended:
+                recommended.append(movie)
+    return recommended
+
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+def get_new_rec_by_genre(user_data):
+    recommended = []
+    if not len(user_data["watched"]):
+        return recommended
+    
+    genres = [movie["genre"] for movie in user_data["watched"]]
+    genre = mode(genres)
+    friends_watched = get_friends_unique_watched(user_data)
+    for movie in friends_watched:
+        if movie["genre"] == genre and movie not in user_data["watched"] and movie not in recommended:

```

Comment:
`get_friends_unique_watched` returns a list of movies that friends have watched but the user has not watched so we can drop the check `movie not in user_data["watched"]` here. Also if friends_watched doesn't have any duplicate movies in it, could we assume that the same movie won't get appended to `recommended` more than once? If we can make that assumption, then we could drop the check for `movie not in recommended` too.

I noticed that line 96 is longer than what Python style guide recommends. If you were to have all 3 checks on line 96, consider storing the return value of the expressions in descriptively named variable to help shorten this line

```python
is_not_in_user_data = movie not in user_data["watched"]
is_not_in_recommended = movie not in recommended
if movie["genre"] == genre and is_not_in_user_data and is_not_in_recommended:
```

It also increases readability.